posthog/migrations/0389_persondistinctid_nullable_person_id.py

```diff
@@ -0,0 +1,28 @@
+# Generated by Django 3.2.23 on 2024-02-06 16:50
+
+from django.db import migrations, models
+from posthog.models.utils import WARN_ON_CASCADE
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("posthog", "0388_add_schema_to_batch_exports"),
+    ]
+
+    operations = [
+        migrations.RunSQL(
+            "ALTER TABLE posthog_persondistinctid ALTER COLUMN person_id DROP NOT NULL",
+            # XXX: This reverse migration really only exists for backwards
+            # compatibility within tests and local development: ``SET NOT NULL``
+            # would be far too expensive to apply on this column in production,
+            # (and liable to fail anyway if a ``NULL`` value has been written.)
+            reverse_sql="ALTER TABLE posthog_persondistinctid ALTER COLUMN person_id SET NOT NULL",
+            state_operations=[
+                migrations.AlterField(
+                    model_name="persondistinctid",
+                    name="person",
+                    field=models.ForeignKey(null=True, on_delete=WARN_ON_CASCADE, to="posthog.person"),
+                ),
+            ],
+        )
```

Comment:
The autogenerated migration:

```sql
BEGIN;
--
-- Alter field person on persondistinctid
--
SET CONSTRAINTS "posthog_persondistin_person_id_5d655bba_fk_posthog_p" IMMEDIATE;
ALTER TABLE "posthog_persondistinctid" DROP CONSTRAINT "posthog_persondistin_person_id_5d655bba_fk_posthog_p";
ALTER TABLE "posthog_persondistinctid" ALTER COLUMN "person_id" DROP NOT NULL;
ALTER TABLE "posthog_persondistinctid" ADD CONSTRAINT "posthog_persondistin_person_id_5d655bba_fk_posthog_p" FOREIGN KEY ("person_id") REFERENCES "posthog_person" ("id") DEFERRABLE INITIALLY DEFERRED;
COMMIT;`
```

I'm not sure why Django thought it was necessary to drop and re-add the constraint here, because the migration works fine without doing so. I was little concerned that would cause Postgres to try and revalidate the integrity of the foreign key constraint on re-creation, which would be relatively slow on such a large table and likely require additional locks on `posthog_person`, so I changed it to leave the constraint as-is and just change the column type.

The locks taken during the migration:

```sql
SELECT
        pg_locks.relation::regclass,
        pg_locks.*
FROM pg_stat_activity
LEFT JOIN pg_locks ON pg_locks.pid = pg_stat_activity.pid
WHERE
        pg_stat_activity.query LIKE '%ALTER TABLE%'
        AND pg_stat_activity.pid != pg_backend_pid()
```

```
         relation         |   locktype    | database | relation | page | tuple | virtualxid | transactionid | classid | objid | objsubid | virtualtransaction |  pid  |        mode         | granted | fastpath 
--------------------------+---------------+----------+----------+------+-------+------------+---------------+---------+-------+----------+--------------------+-------+---------------------+---------+----------
                          | virtualxid    |          |          |      |       | 10/14378   |               |         |       |          | 10/14378           | 87503 | ExclusiveLock       | t       | t
                          | transactionid |          |          |      |       |            |         81190 |         |       |          | 10/14378           | 87503 | ExclusiveLock       | t       | f
 posthog_persondistinctid | relation      |    16384 |    17314 |      |       |            |               |         |       |          | 10/14378           | 87503 | AccessExclusiveLock | t       | f
(3 rows)
```

This _should be_ brief, as altering the column shouldn't be particularly expensive (no constraint validation, type modifications, etc.)

Reply:
Could django be re-creating the fk constraint to drop the `MATCH FULL` clause from it now that the underlying column has become nullable? Anyways, may be worth it to peak at the fk definition.